### PR TITLE
Additional subnets

### PR DESCRIPTION
**Pull request by atsikham**
Current solution: place subnets to available AZs in round-robin, place VMs to created subnets in round-robin. This way we achieve possibility to have VMs in different AZs. Another approach can be applied in future tasks when requirements will be clear and it will be defined whether we attach VMs to subnets or vise-versa (probably #27 issue).

[Already discussed] Points to discuss:
* Input `subnets` parameter format
* Does it meet #20 task requirements? Now the number of nat gateways can be specified and subnets are created in different availability zones using Terraform's `element()` function.
* Should networking part be removed from AWSKS module? May be leave it there as is as now there is an ability to pass existing subnets. It probably will affect tests implementation there.

### CODE REVIEW

**Comment by mkyc** [comment]
PR #32 has to be merged before this PR.

**Comment by mkyc** [comment]
Regarding draft PR description: 
 - Comment from @pprach would be helpful: do we need to specify subnets ranges manually or is that ok to calculate them automatically (at least for now). 
 - but I think, that moving `vms_count` to subnets (to indicate how many machines we want in each subnet) would be required

I'm not sure about #20 and networking part in AWSKS, that would require some discussion.

**Comment by atsikham** [comment]
> Regarding draft PR description:
> 
> * Comment from @pprach would be helpful: do we need to specify subnets ranges manually or is that ok to calculate them automatically (at least for now).
> * but I think, that moving `vms_count` to subnets (to indicate how many machines we want in each subnet) would be required
> 
> I'm not sure about #20 and networking part in AWSKS, that would require some discussion.

* I would calculate CIDR ranges automatically unless we have no such requirement.
* But we are also going to implement [VM groups](https://github.com/epiphany-platform/m-aws-basic-infrastructure/issues/27#issuecomment-728776603), where subnets will be specified. In this case I think your second point conflicts with that task.

**Comment by mkyc** [comment]
is `count` in:
```
 private: [
    {count: `num`, availability_zone: `az_name` or `any`, net_bits: `num`},
    ...
  ],
```

number of machines in this private subnet, or number of such subnets? Can you rename that variable to something more descriptive?

**Comment by atsikham** [comment]
@mkyc modified, it's the number of subnets with the same AZ and [newbits](https://registry.terraform.io/modules/hashicorp/subnets/cidr/latest#new_bits) parameters

**Comment by mkyc** [comment]
How will I define how many machines I want in each subnet?

**Comment by atsikham** [comment]
> How will I define how many machines I want in each subnet?

This is defined by the third parameter in a map (`subnet_address_bits` for example). Assuming you have `VPC CIDR block` defined to `10.1.0.0/20` and `subnet_address_bits` `4`, so there are 8 bits left for hosts and in case of AWS it's 251 host.

**Comment by atsikham** [comment]
> I meant `M_VMS_COUNT` parameter described [here](https://github.com/epiphany-platform/m-aws-basic-infrastructure/blob/develop/docs/INPUTS.adoc). So how will I force AwsBI module to put machines in those new subnets?

At the moment I placed them to public/private subnets depending on `use_public_ip` parameter, with round-robin assignement. I think this sould be implemented in #27 the way @sk4zuzu suggested. But in general I see what you are talking about - we need some identifier to place VM into subnet. But isn't it enough to specify AZ to place vm in and select any subnet from there?

**Comment by mkyc** [comment]
As far as I understand we are talking about direction of relation (vm group in subnet or subnet in vm group). My opinion is that we want to put vm groups in subnets. 

In @sk4zuzu [comment](https://github.com/epiphany-platform/m-aws-basic-infrastructure/issues/27#issuecomment-728776603):

```
define subnet(s) for the group, but it should be also possible to span multiple groups across single subnet 
```

I tend to focus more on `span multiple groups across single subnet` part of sentence. That indicates rather vm groups in subnets. 


BTW: do you consider move of that "module" code to root terraform directory? It feels like overhead.

**Comment by atsikham** [comment]
This request is blocked by #32, but let's review and make changes when new requirements appear.

**Comment by mkyc** [comment]
Let's have this move in separate task, not to overcomplicate this one.

**Comment by mkyc** [comment]
@atsikham please rerun tests as #41 was merged.

File: tests/integration_test.go [hunk 1]
```diff
@@ -370,12 +370,11 @@ func runDocker(t *testing.T, params ...string) (bytes.Buffer, bytes.Buffer) {
 	}
 
 	if err := command.Run(); err != nil {
-		t.Fatal("There was an error running command:", err)
+	    t.Log("Stdout: ", string(stdout.Bytes()))
```
Comment (erzetpe) [inline tests/integration_test.go:373]: At least this should be taken out of this if section.

**Review by erzetpe — CHANGES_REQUESTED** [review]

**Review by erzetpe — APPROVED** [review]

**Comment by mkyc** [comment]
@sk4zuzu can you also confirm that this PR solves #20?

File: resources/terraform/modules/ec2/variables.tf [hunk 2]
```diff
@@ -45,14 +45,20 @@ variable "vpc_cidr_block" {
...
+  type = object({
+    private = object({
+      count = number
+    })
+    public = object({
+      count = number
+    })
+  })
+  validation {
+    condition     = var.subnets.private.count > 0 || var.subnets.public.count > 0
```
Comment (sk4zuzu) [inline resources/terraform/modules/ec2/variables.tf:59]: Does it make sense to have only private subnets without Internet access? What would be the use case? :thinking: 
```suggestion
    condition     = (var.subnets.private.count > 0 && var.subnets.public.count > 0) || var.subnets.public.count > 0
```

**Review by sk4zuzu — DISMISSED** [review]
Looks correct. Clock arithmetic is in place. If there are less NGs than AZs all remaining are reused correctly. I believe this is the HA (#20) that we want. :+1:

File: resources/terraform/modules/ec2/variables.tf (hunk 2 above)
Comment (atsikham) [reply resources/terraform/modules/ec2/variables.tf:59]: Discussed with @sk4zuzu that having private subnet implies that you need public one as well.

**Review by sk4zuzu — APPROVED** [review]

**Review by mkyc — DISMISSED** [review]

File: docs/INPUTS.adoc [hunk 3]
```diff
@@ -14,8 +14,24 @@
...
+  private: {
+    count: 1
+  },
+  public: {
+    count: 1
+  }
+}
+----
+|no |init |Defines number of public and private subnets
+that a created in available AZs with round-robin
```
Comment (sk4zuzu) [inline docs/INPUTS.adoc:34]: Is this a typo? :thinking: 
```suggestion
that are created in available AZs with round-robin
```

**Review by erzetpe — APPROVED** [review]

**Review by sk4zuzu — APPROVED** [review]
:monkey:

**Review by mkyc — APPROVED** [review]
LGTM

**Pull request merged**